privacyidea/lib/applications/offline.py

```diff
@@ -101,6 +53,81 @@ class MachineApplication(MachineApplicationBase):
     """
     application_name = "offline"
 
+    @staticmethod
+    def generate_new_refilltoken(token_obj):
+        """
+        Generate new refill token and store it in the tokeninfo of the token.
+        :param token_obj: token in question
+        :return: a string
+        """
+        new_refilltoken = geturandom(REFILLTOKEN_LENGTH, hex=True)
+        token_obj.add_tokeninfo("refilltoken", new_refilltoken)
+        return new_refilltoken
+
+    @staticmethod
+    def get_offline_otps(token_obj, otppin, amount, rounds=ROUNDS):
+        """
+        Retrieve the desired number of passwords (= PIN + OTP), hash them
+        and return them in a dictionary. Increase the token counter.
+        :param token_obj: token in question
+        :param otppin: The OTP PIN to prepend in the passwords. The PIN is not validated!
+        :param amount: Number of OTP values (non-negative!)
+        :param rounds: Number of PBKDF2 rounds
+        :return: dictionary
+        """
+        if amount < 0:
+            raise ParameterError("Invalid refill amount: {!r}".format(amount))
+        (res, err, otp_dict) = token_obj.get_multi_otp(count=amount, counter_index=True)
+        otps = otp_dict.get("otp")
+        for key in otps.keys():
+            # Return the hash of OTP PIN and OTP values
+            otps[key] = passlib.hash. \
+                pbkdf2_sha512.encrypt(otppin + otps.get(key),
+                                      rounds=rounds,
+                                      salt_size=10)
+        # We do not disable the token, so if all offline OTP values
+        # are used, the token can be used the authenticate online again.
+        # token_obj.enable(False)
+        # increase the counter by the consumed values and
+        # also store it in tokeninfo.
+        token_obj.inc_otp_counter(increment=amount)
+
+        return otps
+
+    @staticmethod
+    def get_refill(token_obj, password, options=None):
+        """
+        Returns new authentication OTPs to refill the client
+
+        To do so we also verify the password, which may consist of PIN + OTP.
+
+        :param token_obj: Token object
+        :param password: PIN + OTP
+        :param options: dict that might contain "count" and "rounds"
+        :return: a dictionary of auth items
+        """
+        options = options or {}
+        count = int(options.get("count", 100))
+        rounds = int(options.get("rounds", ROUNDS))
+        _r, otppin, otpval = token_obj.split_pin_pass(password)
+        if not _r:
+            raise ParameterError("Could not split password")
+        current_token_counter = token_obj.token.count
+        first_old_counter = current_token_counter - count
+        if first_old_counter < 0:
```

Comment:
@cornelinux I'm wondering if this is even a legal case? In case of a refill, shouldn't the current token counter value always be higher than the refill amount?

Reply:
I think you are right. But I am not sure here. Because in some tests I ran into this problem.

    if first_old_counter < 0:
         log.error("This should not happen")  

;-)

And I think you said: If we have a negative value, the counter on the server would be decreased. So this might be the safe net for states, that should not occur!

Reply:
Ah, I guess this could happen if a token gets 100 offline OTPs (i.e. the online token counter is 101), then the administrator increases the offline OTP count to 200 and the client attempts a refill (i.e. ``101 - 200 = -99``).

But apart from that, we ensure that we never decrease the counter on the server by checking for ``amount < 0`` in ``get_offline_otps``.